### PR TITLE
Fix for issue #14.

### PR DESCRIPTION
**Pull request by mkuculyma**
Problem that occurred was that for some kind of metrics facter tool returned nested map instead of simple type value (string, float). So the response could not be encoded in here https://github.com/intelsdi-x/snap/blob/8539f5ac78cc75b7b1e6b8cc31934ce48c4bd534/control/plugin/collector_proxy.go#L99 without prior registration. 

The implemented fix solves this issue by parsing map returned by facter into several different metrics.
i.e. When requested metric is "/intel/facter/os", facter returns "map:[release:map[major:7 minor:2 full:7.2.1511] name:CentOS family:RedHat]".
This value is then parsed into several metrics:
/intel/facter/os/name CentOS
/intel/facter/os/family RedHat
/intel/facter/os/release/major 7
/intel/facter/os/release/minor 2
/intel/facter/os/release/full 7.2.1511

**Note: this fix uses updated snap-plugin-utilities GetValueByNamespace function.**

Changes in code:
- Implementation of solution - when facter returns map instead of a value, the map is being parsed into several metrics.
- Unit tests updated
- License added
- Version bumped
- Godeps updated

### CODE REVIEW

File: facter/facter.go [hunk 1]
```diff
@@ -197,6 +209,28 @@ func (f *Facter) GetConfigPolicy() (*cpolicy.ConfigPolicy, error) {
...
+
+	namespaces := []string{}
+
+	err := utils.FromMap(value, name, &namespaces)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, namespace := range namespaces {
+		ns := strings.Split(namespace, "/")
```
Comment (lmroz) [inline facter/facter.go]: validateNamespace expects namespace to contain exactly 3 segments, GetMetricTypes return such namespaces but here you are expecting additional elements of namespace. They cannot be collected if not reported by GetMetricTypes.

Comment (mkuculyma) [inline facter/facter.go]: fixed

**Comment by lmroz** [comment]
Just skimmed through the code so maybe I'm missing something but taking the example:
```
/intel/facter/os/name CentOS
/intel/facter/os/family RedHat
/intel/facter/os/release/major 7
/intel/facter/os/release/minor 2
/intel/facter/os/release/full 7.2.1511
```

`CollectMetrics` extracts `ns[2]` element from namespace which is in the example `"os"` five times. So maybe you should use some kind of a Set to get rid of duplicates?

File: facter/facter.go [hunk 2]
```diff
@@ -169,18 +165,16 @@ func (f *Facter) CollectMetrics(metricTypes []plugin.MetricType) ([]plugin.Metri
 		return nil, err
 	}
 
-	// make sure that recevied len of names equals asked
-	if len(facts) != len(names) {
-		return nil, errors.New("assertion: getFacts returns more/less than asked!")
+	// make sure that facts has returned sth
+	if len(facts) == 0 {
```
Comment (lmroz) [inline facter/facter.go]: If you get rid of duplicates this assertion perhaps should stay as it was.

File: facter/facter.go [hunk 3]
```diff
@@ -197,11 +191,52 @@ func (f *Facter) GetConfigPolicy() (*cpolicy.ConfigPolicy, error) {
...
+		val := utils.GetValueByNamespace(value, ns[1:])
+		m := *plugin.NewMetricType(createNamespace(ns...), time.Now(), nil, "", val)
+		ms = append(ms, m)
+	}
+
+	return ms, nil
+}
+
 // validateNamespace checks namespace intel(vendor)/facter(prefix)/FACTNAME
 func validateNamespace(namespace core.Namespace) error {
```
Comment (lmroz) [inline facter/facter.go]: Perhaps this functionis unnecessary and namespace routing is carried by framework but if you decided to leave it here the check should be changed to `len(namespace) >= 3`.

File: facter/facter.go (hunk 2 above)
Comment (mkuculyma) [inline facter/facter.go]: Fixed

File: facter/facter.go (hunk 3 above)
Comment (mkuculyma) [inline facter/facter.go]: Fixed

File: facter/facter_integration_test.go [hunk 4]
```diff
@@ -76,18 +77,6 @@ func TestFacterCollectMetrics(t *testing.T) {
 		})
 
 		Convey("ask for inappriopriate metrics", func() {
-			Convey("wrong number of parts", func() {
```
Comment (katarzyna-z) [inline facter/facter_integration_test.go]: Why did you remove this part?
